torch_np/_ndarray.py

```diff
@@ -147,11 +223,105 @@ def ravel(self, order='C'):
             raise NotImplementedError
         return ndarray._from_tensor_and_base(self._tensor.ravel(), self)
 
+    def nonzero(self):
+        tensor = self._tensor
+        return tuple(asarray(_) for _ in tensor.nonzero(as_tuple=True))
+
+    @axis_out_keepdims_wrapper
+    def any(self, axis=None, out=None, keepdims=NoValue, *, where=NoValue):
+        if where is not None:
+            raise NotImplementedError
+
+        axis = _helpers.allow_only_single_axis(axis)
+
+        if axis is None:
+            result = self._tensor.any()
+        else:
+            result = self._tensor.any(axis)
+        return result
+
+
+    @axis_out_keepdims_wrapper
+    def all(self, axis=None, out=None, keepdims=NoValue, *, where=NoValue):
+        if where is not None:
+            raise NotImplementedError
+
+        axis = _helpers.allow_only_single_axis(axis)
+
+        if axis is None:
+            result = self._tensor.all()
+        else:
+            result = self._tensor.all(axis)
+        return result
+
+
+    @axis_out_keepdims_wrapper
+    def max(self, axis=None, out=None, keepdims=NoValue, initial=NoValue,
+             where=NoValue):
+        if where is not None:
+            raise NotImplementedError
+        if initial is not None:
+            raise NotImplementedError
+
+        result = self._tensor.amax(axis)
+        return result
+
+    @axis_out_keepdims_wrapper
+    def min(self, axis=None, out=None, keepdims=NoValue, initial=NoValue,
+             where=NoValue):
+        if where is not None:
+            raise NotImplementedError
+        if initial is not None:
+            raise NotImplementedError
+
+        result = self._tensor.amin(axis)
+        return result
```

Comment:
@lezcano could you please take a look when you have time? Two points I'm not sure about interactions with pytorch:

- it's convenient to use `torch.amax` here, not `torch.max`, because `amax` supports tuple axes while `max` doesn't; there is however a cryptic comment in https://pytorch.org/docs/stable/generated/torch.amax.html?highlight=amax#torch.amax, _amax/amin evenly distributes gradient between equal values, while max(dim)/min(dim) propagates gradient only to a single index in the source tensor._ --- is this relevant to use here?

- since `dim / keepdim` handling is not entirely uniform across reductions, this PR opts to not propagating `keepdim` to torch at all, and handle it in the `@axis_out_keepdims_wrapper`. Does it play well with dynamo or inductor or who is the consumer of this?

Other than these, I think this PR is a reasonable approximation to generic handling of reductions, so a review would be appreciated if you have time.

Reply:
Always prefer `amax` over `max` because of that point about gradients. This will be relevant once we start considering using this wrapper to compute derivatives of NumPy programs. For reference, the comment references the case when you compute the maximum of a tensor with repeated values. The function `max` is not differentiable at these points, but it's convex, so we compute its subgradient of max norm.

About `dim`/`keepdim`, do you mean that it's not uniform on PyTorch or on NumPy? If it's the former, can you give an example? We should fix it in PyTorch if that's the case.

Reply:
Thanks, will keep using `amin, amax` then. 

dim/keepdim comment is about pytorch:

```
In [33]: torch.any([0, 1, 2, 3], keepdim=True)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [33], in <cell line: 1>()
----> 1 torch.any([0, 1, 2, 3], keepdim=True)

TypeError: any() received an invalid combination of arguments - got (list, keepdim=bool), but expected one of:
 * (Tensor input, *, Tensor out)
 * (Tensor input, int dim, bool keepdim, *, Tensor out)
 * (Tensor input, name dim, bool keepdim, *, Tensor out)
```

but e.g. `amax` accepts keepdim without an explicit axis (but refuses to convert a list into a tensor):

```
In [34]: torch.amax([0, 1, 2, 3], keepdim=True)                    # oops, lists are not accepted
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [34], in <cell line: 1>()
----> 1 torch.amax([0, 1, 2, 3], keepdim=True)

TypeError: amax(): argument 'input' (position 1) must be Tensor, not list

In [35]: torch.amax(torch.as_tensor([0, 1, 2, 3]), keepdim=True)
Out[35]: tensor([3])
```

Reply:
Functions in PyTorch do not accept list in place of tensors, so the second part is expected.

About the first part, yep, `any` is wrong. Now, I don't know if we can fix this in a BC way. WDYT @ezyang @albanD? If it's alright to fix this, I can submit a fix.

For reference, this is the file that binds the C++ API to the Python API. 
https://github.com/pytorch/pytorch/blob/2b52db9c953d063db7b46c12f4df35b47aca4381/aten/src/ATen/native/native_functions.yaml#L9401
In particular, if you look up `amax` and (all the appearances of) `any`, it should be clear what's the issue. For how to read this file, see the `README.md` in that same folder.

Reply:
Looks like a pretty easy, BC fix. We'd just add another overload without an `int dim` argument.

I wouldn't necessarily block work here on it though; keepdim without a dim argument is easy enough to simulate, no?

Reply:
Exactly. This whole thread started with me asking if it's OK from torchinductor perspective to emulate keepdim in this layer (https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/8/files#diff-90a04d87d7a3e373f1e17e5f80c183769c147e7034e66d1b8d6e57efe61cc150R100) and always call pytorch functions without `keepdim`. Now that both Mario and you confirmed it's indeed OK, I happily continue :-).